pkg/k8s/jobrunner.go

```diff
@@ -27,50 +25,71 @@ type jobRunner struct {
 }
 
 func (j *jobRunner) Run(job *batchv1.Job) error {
+	ctx := j.kube.Context()
 	jobs := j.kube.Typed().BatchV1().Jobs(job.Namespace)
-	_, err := jobs.Create(j.kube.Context(), job, metav1.CreateOptions{})
+	_, err := jobs.Create(ctx, job, metav1.CreateOptions{})
 	if err != nil {
-		return fmt.Errorf("%w: %v", ErrUnexcpected, err)
+		return fmt.Errorf("%w: %v", ErrICSenderJobFailed, err)
 	}
-	factory := kubeinformers.NewSharedInformerFactoryWithOptions(
-		j.kube.Typed(),
-		time.Minute,
-		kubeinformers.WithNamespace(job.Namespace),
-		kubeinformers.WithTweakListOptions(func(options *metav1.ListOptions) {
-			options.FieldSelector = fmt.Sprintf("metadata.name=%s", job.Name)
-		}),
-	)
-	// FIXME: This function do not wait properly for the end of the Job
-	stop := make(chan struct{})
-	jobsInformer := factory.Batch().V1().Jobs().Informer()
-	jobsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			close(stop)
-		},
+	err = j.watchJob(job, func(job *batchv1.Job) (bool, error) {
+		if job.Status.CompletionTime == nil && job.Status.Failed == 0 {
+			return false, nil
+		}
+		// We should be done if we reach here.
+		if job.Status.Succeeded < 1 {
+			return false, fmt.Errorf("%w: %s", ErrICSenderJobFailed,
+				"expected to have successful job")
+		}
+		return true, nil
 	})
-	go factory.Start(stop)
-	waitOnStop(stop)
-	updated, err := jobs.Get(j.kube.Context(), job.Name, metav1.GetOptions{})
 	if err != nil {
-		return fmt.Errorf("%w: %v", ErrUnexcpected, err)
+		return fmt.Errorf("%w: %v", ErrICSenderJobFailed, err)
 	}
-	if updated.Status.Succeeded < 1 {
-		return fmt.Errorf("%w: %s", ErrUnexcpected, "expected to have successful job")
+	return j.deleteJob(job)
+}
+
+func (j *jobRunner) deleteJob(job *batchv1.Job) error {
+	ctx := j.kube.Context()
+	jobs := j.kube.Typed().BatchV1().Jobs(job.GetNamespace())
+	err := jobs.Delete(ctx, job.GetName(), metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrICSenderJobFailed, err)
 	}
-	err = jobs.Delete(j.kube.Context(), job.Name, metav1.DeleteOptions{})
+	pods := j.kube.Typed().CoreV1().Pods(job.GetNamespace())
+	err = pods.DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("job-name=%s", job.GetName()),
+	})
 	if err != nil {
-		return fmt.Errorf("%w: %v", ErrUnexcpected, err)
+		return fmt.Errorf("%w: %v", ErrICSenderJobFailed, err)
 	}
 	return nil
 }
 
-func waitOnStop(stop chan struct{}) {
-	for {
-		select {
-		case <-stop:
-			return
-		default:
-			time.Sleep(time.Second)
+func (j *jobRunner) watchJob(meta metav1.Object, changeFn func(job *batchv1.Job) (bool, error)) error {
+	ctx := j.kube.Context()
+	jobs := j.kube.Typed().BatchV1().Jobs(meta.GetNamespace())
+	watcher, err := jobs.Watch(ctx, metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", meta.GetName()),
+	})
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrICSenderJobFailed, err)
+	}
```

Comment:
Would be `defer watcher.Stop()` helpful around here to clean up resources?

Reply:
And if check `RetryWatcher` from `client-go` tools. In that case it's trying to restart on closed `ResultChan` channel. That's more likely a rare case but we've seen such reports in client's issue that `ResultChan` is closed unexpectedly, but with recoverable error.

Finally I wonder if you couldn't reuse something like existing RetryWatcher code directly here.

https://github.com/kubernetes/client-go/blob/master/tools/watch/retrywatcher.go#L242-L275

Reply:
I'm thinking that maybe that's not be possible in this exact case because of:

https://github.com/kubernetes/client-go/blob/ac207faedfb64acd5b99a2fb309b7044918b4dda/tools/watch/retrywatcher.go#L68-L71

WDYT?

Reply:
Ahh, I recall that part now. Sure let's keep your watching implementation.

Do you think it's worth to address at least a bit of retry logic when `ResultChan` is closed? Of course it can done as a future enhancement/hardening.

Reply:
I think it's good idea for an enhancement in future PR. Would you mind opening an issue?

Reply:
Yep, I'll do first thing tomorrow.